plugins/a11y-keyboard/msd-a11y-keyboard-atspi.c

```diff
@@ -0,0 +1,193 @@
+/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*-
+ *
+ * Copyright (C) 2020 Colomban Wendling <cwendling@hypra.fr>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "config.h"
+
+#include "msd-a11y-keyboard-atspi.h"
+
+#include <glib-object.h>
+
+#if HAVE_LIBATSPI
+#include <gdk/gdk.h>
+#include <atspi/atspi.h>
+
+/* See https://gitlab.gnome.org/GNOME/at-spi2-core/-/issues/22 */
+#if defined (HAVE_LIBATSPI_NOBUG22) && HAVE_LIBATSPI_NOBUG22
+#       define DESTROYING_ATSPI_LISTENER_CRASHES 0
+#else
+#       define DESTROYING_ATSPI_LISTENER_CRASHES 1
+#endif
+
+
+struct MsdA11yKeyboardAtspiPrivate
+{
+        AtspiDeviceListener *listener;
+        gboolean             listening;
+};
+
+G_DEFINE_TYPE_WITH_PRIVATE (MsdA11yKeyboardAtspi, msd_a11y_keyboard_atspi, G_TYPE_OBJECT)
+
+static void
+msd_a11y_keyboard_atspi_finalize (GObject *obj)
+{
+        MsdA11yKeyboardAtspi *self = MSD_A11Y_KEYBOARD_ATSPI (obj);
+
+        g_clear_object (&self->priv->listener);
+        self->priv->listening = FALSE;
+
+        G_OBJECT_CLASS (msd_a11y_keyboard_atspi_parent_class)->finalize (obj);
+}
+
+static void
+msd_a11y_keyboard_atspi_class_init (MsdA11yKeyboardAtspiClass *klass)
+{
+        GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+        object_class->finalize = msd_a11y_keyboard_atspi_finalize;
+}
+
+static gboolean
+on_key_press_event (const AtspiDeviceEvent *event,
+                    void                   *user_data G_GNUC_UNUSED)
```

Comment:
IMHO, it's not a good idea to silence the warnings.

Reply:
LGTM this change

Reply:
@rbuj why?  IMO `-Wunused-parameter` is a fair warning that can help spot bugs and suggest refactoring, but here as it's a callback it has a specific signature I can't alter, yet I really don't need it here, as I know it's value (it's useful when the signal can be triggered for several GSettings keys, but here it's set up to only be emitted for a specific one, so it's pointless to check again).
So IMO it's a good idea to silence this instance of the warning as it's a kind of false positive, and would make the other warnings harder to spot in the mix.

Reply:
If you want to filter out this warning type, you can use "-Wno-unused-variable", if you don't need a parameter you can use the function cast macro G_CALLBACK or use void for overriding the parameter within the function body, but using G_GNUC_UNUSED silents this warnings, which cannot be detected if you add G_GNUC_UNUSED.

Reply:
G_GNUC_UNUSED only appears 27 times:
```
$ grep -R G_GNUC_UNUSED --include=*.c
mate-session-manager/mate-session/test-inhibit.c:        const char *startup_id G_GNUC_UNUSED;
marco/src/ui/tabpopup.c:                             GdkScreen *old_screen G_GNUC_UNUSED,
marco/src/ui/tabpopup.c:                             gpointer   data G_GNUC_UNUSED)
marco/src/core/main.c:      G_GNUC_UNUSED int dummy;
libmatemixer/backends/oss/oss-backend.c:                                                  guint             index) G_GNUC_UNUSED;
mate-terminal/src/terminal-screen-container.c:                                      GtkPolicyType hpolicy G_GNUC_UNUSED,
engrampa/src/fr-init.c:G_GNUC_UNUSED static void
engrampa/src/fr-init.c:G_GNUC_UNUSED static gboolean
engrampa/src/eggfileformatchooser.c:                             gpointer value G_GNUC_UNUSED,
engrampa/src/eggfileformatchooser.c:                GtkTreePath  *path G_GNUC_UNUSED,
engrampa/src/eggfileformatchooser.c:                   GtkTreePath  *path G_GNUC_UNUSED,
engrampa/src/eggfileformatchooser.c:                                            guint                 parent G_GNUC_UNUSED,
caja/src/caja-application.c:                  GParamSpec     *unused G_GNUC_UNUSED,
pluma/pluma/pluma-print-preview.c:dummy_write_func (G_GNUC_UNUSED gpointer      closure,
pluma/pluma/pluma-print-preview.c:		  G_GNUC_UNUSED const guchar *data,
pluma/pluma/pluma-print-preview.c:		  G_GNUC_UNUSED guint         length)
mate-panel/applets/notification_area/testtray.c:signal_handler (gpointer data G_GNUC_UNUSED)
mate-panel/mate-panel/panel-test-applets.c:G_GNUC_UNUSED void on_execute_button_clicked (GtkButton *button, gpointer dummy);
mate-panel/mate-panel/panel-test-applets.c:G_GNUC_UNUSED void
mate-settings-daemon/plugins/mouse/msd-mouse-manager.c:        G_GNUC_UNUSED XEventClass class_presence;
mate-settings-daemon/plugins/mouse/msd-mouse-manager.c:        G_GNUC_UNUSED  int xi_presence;
mate-settings-daemon/plugins/xsettings/fontconfig-monitor.c:stuff_changed (GFileMonitor *monitor G_GNUC_UNUSED,
mate-settings-daemon/plugins/xsettings/fontconfig-monitor.c:               GFile *file G_GNUC_UNUSED,
mate-settings-daemon/plugins/xsettings/fontconfig-monitor.c:               GFile *other_file G_GNUC_UNUSED,
mate-settings-daemon/plugins/xsettings/fontconfig-monitor.c:               GFileMonitorEvent event_type G_GNUC_UNUSED,
mate-settings-daemon/plugins/a11y-keyboard/msd-a11y-keyboard-manager.c:        G_GNUC_UNUSED XEventClass class_presence;
mate-settings-daemon/plugins/a11y-keyboard/msd-a11y-keyboard-manager.c:        G_GNUC_UNUSED int xi_presence;
```

Reply:
> If you want to filter out this warning type, you can use "-Wno-unused-parameter",

Yes, but that will filter useful warnings as well.  The point of marking the ones that are legitimate is to make it clear to everyone it's known and OK, and not to clutter the compiler output.  It's basically impossible to spot 1 useful warning if you have a thousand false positive, so the useful one would just get lost.

> if you don't need a parameter you can use the function cast macro G_CALLBACK

Yes, in this case it would work.  It's slightly wrong on paper as it still is calling an incorrectly typed function, but in practice it'll work as GTK itself uses this kind of trick so all architectures it runs on ought to be OK with this.

> or use void for overriding the parameter within the function body, but using G_GNUC_UNUSED silents these warnings, which cannot be detected if you add G_GNUC_UNUSED.

I don't get this. What's the difference between making dummy use of the argument or marking it G_GNUC_UNUSED?  It even seems easier to me to spot G_GNUC_UNUSED than `(void) param;`, so I'd expect it to be *easier* to detect (whatever you mean by that).

> G_GNUC_UNUSED only appears 27 times in about 665000 lines of code:
>
> without including the 27 muted warnings there are 6776 warnings about -Wunused-parameter in our current build logs

What point are you trying to make?  If it's that G_GNUC_UNUSED is not used much in the codebase, OK, but that doesn't necessarily mean it *should* not be.  And if you tell me there are 6776 warnings about this, I wonder if we indeed shouldn't start doing something about this (not saying it's a priority, but still).
I don't see your data as suggesting we should not be using any mean of silencing unused parameter warnings, rather the contrary.  Which technique to use is another question, and if there's a consensus on that I'm happy to follow it, but for the moment I still like G_GNUC_UNUSED.

Reply:
I agree 100% with @cwendling 

@cwendling maybe you like to review https://github.com/mate-desktop/mate-utils/pull/300 which can be handled easily with G_GNUC_UNUSED :)

Reply:
What I’m pointing you is that the codebase prefers to show all warnings instead of cheating and hiding the ones you don't like to show.

Once we have reached this point in the conversation, I think the best is to seek the opinion of other team members. We have both put forward enough arguments for and against the use of the `G_GNUC_UNUSED` macro which is expanded to `unused`.
- https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-unused-function-attribute
- https://clang.llvm.org/docs/AttributeReference.html#maybe-unused-unused

I will respect the decision made by the team.

@mate-desktop/core-team Can you help us decide what is the best option in this case?

Reply:
@rbuj 

> What I’m pointing you is that the codebase prefers to show all warnings instead of cheating and hiding the ones you don't like to show.

it is false positive warning, I don't like to see false positive warnings

at my point of view, the author of PR can decide

> I will respect the decision made by the team.

by the moment, 2 members versus 1

Reply:
I don't know enough about this with the warnings to hold a useful opinion